File: JAVA/Flatten Binary Tree to Linked List.java
public class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode n = root;
        TreeNode last = null;
        while (n != null || !stack.empty()) {
            if (n != null) {
                if (last != null) {
                    last.left = null;
                    last.right = n; 
                }
                last = n;
                if (n.right != null) {
                    stack.push(n.right);
                }
                n = n.left;
            }
            else {
                n = stack.pop();
            }
        }
    }
}

File: Flatten Binary Tree to Linked List.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode *root) {
        TreeNode *tail;
        flattenRe(root);
    }
    
    // return tail
    TreeNode* flattenRe(TreeNode *head) {
        if (head == NULL) {
            return NULL;
        }
        if (head->left == NULL && head->right == NULL) {
            return head;
        }
        TreeNode *leftTail = flattenRe(head->left);
        TreeNode *rightTail = flattenRe(head->right);
        TreeNode *tail;
        TreeNode *rightHead = head->right;
        if (head->left != NULL) {
            head->right = head->left;
            head->left = NULL;
            leftTail->right = rightHead;
            leftTail->left = NULL;
            tail = leftTail;
        }
        if (rightHead != NULL) {
            tail = rightTail;
        }
        return tail;
    }
};
